server/src/main/java/com/paypal/selion/proxy/SeLionSauceProxy.java

```diff
@@ -22,72 +22,248 @@
         }
 
         try {
+            sauceConfigReader = SauceConfigReader.getInstance();
+            defaultParentTunnel = sauceConfigReader.getDefaultParentTunnel();
+            defaultTunnelIdentifier = sauceConfigReader.getDefaultTunnelIdentifier();
             sauceApi = new SauceLabsRestApi();
-        } catch (GridConfigurationException e ) {
+        } catch (GridConfigurationException e) {
             throw new GridException("Failed to initialize proxy: ", e);
         }
     }
 
     @Override
-    public TestSession getNewSession(Map<String, Object> requestedCapability) {
+    public TestSession getNewSession(Map<String, Object> requestedCapabilities) {
+        LOGGER.entering();
         TestSession session = null;
         try {
             accessLock.lock();
 
             try {
+                // sleep for 3 seconds to mitigate hitting sauce REST api rate limits
                 Thread.sleep(3000);
             } catch (InterruptedException e) {
                 LOGGER.log(Level.SEVERE, e.getMessage(), e);
             }
-            if (sauceApi.getNumberOfTCRunning() <= getMaxNumberOfConcurrentTestSessions()) {
-                String username = (String) requestedCapability.get("sauceUserName");
-                String accessKey = (String) requestedCapability.get("sauceApiKey");
-                String tunnelId = (String) requestedCapability.get("parent-tunnel");
-                if (StringUtils.isNotEmpty(username) && StringUtils.isNotEmpty(accessKey)) {
-                    requestedCapability.put("username", username);
-                    requestedCapability.put("accessKey", accessKey);
-                }
-                if (StringUtils.isNotEmpty(tunnelId)) {
-                    requestedCapability.put("parent-tunnel", tunnelId);
-                }
-                session = super.getNewSession(requestedCapability);
+
+            // check for backwards compatible capability names that might come across the wire
+            // TODO remove backwards compatible support in a future version.
+            processDeprecatedCapabilities(requestedCapabilities);
+
+            // default the credentials, if required
+            processCredentails(requestedCapabilities);
+
+            // setup the tunnelIdentifier, if required
+            processTunnelIdentifierCapability(requestedCapabilities);
+
+            // setup the parent-tunnel, if required
+            processParentTunnelCapability(requestedCapabilities);
+
+            // override any idleTimeout capability that may have been specified. Sauce and this proxy need to have
+            // a similar idle timeout value.
+            requestedCapabilities.put(SauceLabsCapability.IDLE_TIMEOUT_CAPABILITY, getTimeOut() / 1000);
+
+            // if call to geNumberOfTCRunning returns -1, sauce REST API calls are failing.
+            // so we kick the request back to the hub.
```

Comment:
The 3 seconds, I believe (not my code) is to mitigate the sauce rest api rate limits -- https://wiki.saucelabs.com/display/DOCS/Rate+Limits+for+the+Sauce+Labs+REST+API

I opted to kick the request back to the hub here for the same reasons we do when `numberOfTestCasesRunning` == `maxConcurrent` ... To give the request another chance up until it is timed out by the `newSessionWaitTimeout` in the hub's config. If sauce it truly down and we honor the session, it will result in a more cryptic WebDriverException error for the client. I'm trying to avoid this too.